### PR TITLE
Entity cache invalidation documentation

### PR DESCRIPTION
**Pull request by Geal**
*Description here*

Fixes #**issue_number**

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:133]:
```suggestion
### Entity cache invalidation
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:148]:
```suggestion
You can configure entity cache invalidation both globally (`preview_entity_cache.subgraph.all.invalidation`) and per subgraph (`preview_entity_cache.subgraph.subgraphs.invalidation`). 
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:135]:
```suggestion
When existing cache entries need to be replaced, the router supports a couple ways for you to invalidate entity cache entries:
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:137]:
```suggestion
- **Invalidation endpoint** - the router exposes an invalidation that can receive invalidation requests from any authorized service out-of-band. A subgraph can use this endpoint to asynchronously trigger invalidation events, or a platform operator can use it to maintain the caching levels.
- **Subgraph response extensions** - you can send invalidation requests via subgraph response extensions, allowing a subgraph to invalidate cached data right after a mutation.
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:142]:
```suggestion
One invalidation request can invalidate multiple cached entries at once. It can invalidate:
- All cached entries for a specific subgraph
- All cached entries for a specific type in a specific subgraph
- All cached entries for a specific entity in a specific subgraph
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:144]:
```suggestion
To process an invalidation request, the router first sends a `SCAN` command to Redis to find all the keys that match the invalidation request. After iterating over the scan cursor, the router sends a `DEL` command to Redis to remove the matching keys.
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:213]:
```suggestion
<Note>

This is the same key format used by the router in the `representations` field for `_entities` queries.

</Note>

```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:217]:
```suggestion
The invalidation endpoint exposed by the router can receive multiple invalidation requests at once and will process them in sequence. For authorization, you can provide a shared key in the request. For example, with the previous configuration you can send the following request:
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:280]:
```suggestion
Entity caching will greatly reduce traffic to subgraphs, resulting in less infrastructure to serve more client traffic. In the rare case where Redis is unavailable, the router should be configured with [rate limiting for subgraph requests](/router/configuration/traffic-shaping/#rate-limiting-1) to avoid overwhelming the subgraphs. It can also be paired with [subgraph query deduplication](/router/configuration/traffic-shaping/#query-deduplication) to further reduce traffic.
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:276]:
```suggestion
- `apollo.router.operations.entity.invalidation.event` - counter triggered when a batch of invalidation requests is received. It has a label `origin` that can be either `endpoint` or `extensions`.
- `apollo.router.operations.entity.invalidation.entry` - counter measuring how many entries are removed per `DEL` call. It has a label `origin` that can be either `endpoint` or `extensions`, and a label `subgraph.name` with the name of the receiving subgraph.
- `apollo.router.cache.invalidation.keys` - histogram measuring the number of keys that were removed from Redis per invalidation request.
- `apollo.router.cache.invalidation.duration` - histogram measuring the time spent handling one invalidation request.

Invalidation requests are also reported under the following spans:
- `cache.invalidation.batch` - span covering the processing of a list of invalidation requests. It has a label `origin` that can be either `endpoint` or `extensions`.
- `cache.invalidation.request` - span covering the processing of a single invalidation request.
```

File: docs/source/configuration/entity-caching.mdx
Comment (chandrikas) [reply docs/source/configuration/entity-caching.mdx:135]: "a couple of ways"

File: docs/source/configuration/entity-caching.mdx
Comment (chandrikas) [reply docs/source/configuration/entity-caching.mdx:137]: Minor correction to Edward's copy

- **Invalidation endpoint** - the router exposes an invalidation **endpoint** that can receive invalidation requests from any authorized service out-of-band. A subgraph can use this endpoint to asynchronously trigger invalidation events, or a platform operator can use it to maintain the caching levels.

File: docs/source/configuration/entity-caching.mdx
Comment (chandrikas) [inline docs/source/configuration/entity-caching.mdx:272]:
```suggestion
- `apollo.router.cache.invalidation.duration`, histogram measuring the time spent handling one invalidation request
```

**Comment by apollo-bot2** [comment]
## Detected Secrets

## :red_circle: Secrets Found

If you are seeing this message, it means that [the secret scanning tool](https://gitleaks.io/) that Apollo uses to prevent secrets from entering our repositories has identified a secret in your branch. Information about what was detected and steps to move forward are below.


If the secret scanner detected a valid secret, please take immediate action to remove it from your branch. Apollo also recommends that you invalidate the secret in the system it grants access to. While there are additional steps to prepare the branch for merging, protecting the secret is the top priority.


If the scanner detected a value that is not actually secret or it detected a value that _was_ a secret but has been invalidated, follow the steps in the "Resolution Process" section below to mark it as a false/benign positive. One common reason for benign positive detections is test values. For example, if you are committing a value (like a private key, JWT, etc.) that would usually be secret but is being used in a test case to validate functionality, a detection will be generated. This is safe to allowlist.

###  Values Detected 

| Signature | File | Commit | Start Line | Start Column | Link To Secret |
|-----------|------|--------|------------|--------------|----------------|
| generic-api-key | docs/source/configuration/entity-caching.mdx | `7443a29fe046c0cd307382e8cf4f0d4c878b8efd` | 166 | 17 | <a href="https://github.com/apollographql/router/blob/7443a29fe046c0cd307382e8cf4f0d4c878b8efd/docs/source/configuration/entity-caching.mdx?plain=1#L166" target="_blank">Link</a> |
| generic-api-key | docs/source/configuration/entity-caching.mdx | `7443a29fe046c0cd307382e8cf4f0d4c878b8efd` | 174 | 19 | <a href="https://github.com/apollographql/router/blob/7443a29fe046c0cd307382e8cf4f0d4c878b8efd/docs/source/configuration/entity-caching.mdx?plain=1#L174" target="_blank">Link</a> |

<hr />
<details>
    <summary><h3>Resolution Process</h3></summary>

> [!NOTE]  
> Apollo does not use all of the native options provided by Gitleaks to perform allowlisting. We use Gitleaks' standard line comments but rely on a custom convention written in a `.gitleaks.toml` file at the root of each repository for allowlisting detections in Git history.

To resolve the issue, you will need to:

1. Prevent new detections of the "secret":
    - Add a comment containing `gitleaks:allow` to the line containing the detected value.
    - Skip this step if it's not feasible. For example, the detection is in file types like JSON or Markdown, you've completely removed the value, etc.
2. Configure the scanner to ignore the "secret" in git history.
    - Follow the instructions below for _either_ "Allowlist by Commit Hash" _or_ "Allowlist by File Path"
    - You will need to ensure that a `.gitleaks.toml` file exists at the root of this repository in your branch. If not, create one!

<details>
    <summary><h4>Allowlist by Commit Hash</h4></summary>

This is the preferred option. Update the `.gitleaks.toml` file with the following contents:

```toml
[[ rules ]]
    id = "signature-from-table-above"
    [ rules.allowlist ]
        commits = [
            "full-commit-hash-1",
            "full-commit-hash-2"
        ]
```

If a `[[rules]]` block already exists for the signature involved in your detection, simply append your list of commits.

If a matching `[[rules]]` block is already present but has a `paths` attribute rather than `commits`, proceed to the next option.

This configuration tells the scanner to ignore the specified signature in the list of commits provided.

</details>
<details>
    <summary><h4>Allowlist by File Path</h4></summary>

This is a backup option to allowlist specific detections. Add/update the `.gitleaks.toml` file with the following contents:

```toml
[[ rules ]]
    id = "signature-from-table-above"
    [ rules.allowlist ]
        paths = [
            '''path/to/detected/file1.json$''',
            '''path/to/other/detected/file.md$'''
        ]
```

This configuration tells the scanner to ignore any detections with the specified signature generated for the list of files.

</details>
</details>
<hr />
<details> 
    <summary><h3>Getting Help</h3></summary>



The Apollo Security team has been notified and is available to assist in resolving this issue. Please tag us on this PR using  `@apollographql/security` if you need assistance! 


</details>
<hr />
<details>
    <summary><h3>How do I know I fixed this correctly?</h3></summary>
Once you have resolved the issue completely, this message will disappear! If you're still seeing this message, there is more to do prior to merging.
</details>

File: docs/source/configuration/entity-caching.mdx
Comment (lleadbet) [inline docs/source/configuration/entity-caching.mdx:215]:
```suggestion
The key field is the same arguments as defined in the subgraph's `@key` directive. If a subgraph has multiple keys defined and the entity is being invalidated, it is likely you'll need to send a request for each key definition. 
```

Comment (lleadbet) [reply docs/source/configuration/entity-caching.mdx:215]: Most subgraph authors do not care or know about the `representations` arg here; it's an implementation detail for their subgraph that most teams have abstracted.

File: docs/source/configuration/entity-caching.mdx
Comment (lleadbet) [inline docs/source/configuration/entity-caching.mdx:222]:
```suggestion
The invalidation endpoint exposed by the router can receive multiple invalidation requests at once and will process them in sequence. 

For authorization, you can provide a shared key in the request. For example, with the previous configuration you can send the following request:
```

File: docs/source/configuration/entity-caching.mdx
Comment (lleadbet) [inline docs/source/configuration/entity-caching.mdx:238]:
```suggestion
The router would send the following response:
```

**Review by lleadbet — APPROVED** [review]

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:136]: In what sense is it asynchronous? The invalidation endpoint waits for the DEL operations to complete and returns a count, which is the number of scanned entries, to the client. That sounds synchronous to me (at least with respect to the client request).

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:179]:
```suggestion
Invalidation requests are defined as JSON objects with the following format:
```
I don't think you can "send" an individual invalidation request to the endpoint. I believe it has to be an array (as documented below).

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:222]:
```suggestion
The invalidation endpoint exposed by the router expects to receive an array of invalidation requests and will process them in sequence. For authorization, you must provide a shared key in the request header. For example, with the previous configuration you should send the following request:
```

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:249]: It's possibly a technicality, but we haven't implemented this. The invalidation `delete` function actually uses the number of matching keys from the scan. The underlying `RedisCacheStorage` `delete` function returns an `Option` which does contain the number of deleted keys, so we could fix that as something like:
```
-                        count += keys.len() as u64;
-                        storage.delete(keys).await;
+                        count += storage.delete(keys).await.unwrap_or(0) as u64;
```

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:285]:
```suggestion
Entity caching will greatly reduce traffic to subgraphs. Should there be an availability issue with a Redis cache, this could cause traffic to subgraphs to increase to a level where infrastructure becomes overwhelmed. To avoid such issues, the router should be configured with [rate limiting for subgraph requests](/router/configuration/traffic-shaping/#rate-limiting-1) to avoid overwhelming the subgraphs. It could also be paired with [subgraph query deduplication](/router/configuration/traffic-shaping/#query-deduplication) to further reduce traffic.
```

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [reply docs/source/configuration/entity-caching.mdx:136]: asynchronous as in it does not have to do that in the middle of handling a request. That can be sent at any point in time

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [reply docs/source/configuration/entity-caching.mdx:215]: good point yes

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [reply docs/source/configuration/entity-caching.mdx:136]: Ah. I find it a bit confusing because the invalidation endpoint itself is synchronous. I'll see if I can think of a better way to describe that.

File: docs/source/configuration/entity-caching.mdx
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:136]: Is this an improvement?

```suggestion
- **Invalidation endpoint** - the router exposes an invalidation endpoint that can receive invalidation requests from any authorized service. This is primarily intended as an alternative to the extensions mechanism described below. For example a subgraph could use it to trigger invalidation events "out of band" from any requests received by the router or a platform operator could use it to invalidate cache entries in response to events which aren't directly related to a router.
```

Comment (Geal) [reply docs/source/configuration/entity-caching.mdx:136]: sounds good

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [inline docs/source/configuration/entity-caching.mdx:175]:
```suggestion

  subgraph:
    all:
      enabled: true
      redis:
        urls: ["redis://..."]
      invalidation:
        # base64 string that will be provided in the `Authorization: Basic` header value
        shared_key: "agm3ipv7egb78dmxzv0gr5q0t5l6qs37"
    subgraphs:
      products:
        # per subgraph invalidation configuration
        invalidation:
          # whether invalidation is enabled for this subgraph
          enabled: true
          # override the shared key for this particular subgraph. If another key is provided, the invalidation requests for this subgraph's entities will not be executed
          shared_key: "czn5qvjylm231m90hu00hgsuayhyhgjv"
```

File: docs/source/configuration/entity-caching.mdx
Comment (lrlna) [inline docs/source/configuration/entity-caching.mdx:136]: can this link to the `Invalidation HTTP Endpoint` header? As I was reading through this, I kept thinking this endpoint is done through regular `router.yaml`, i.e. the section that follows right after this.

File: docs/source/configuration/entity-caching.mdx
Comment (lrlna) [inline docs/source/configuration/entity-caching.mdx:148]: I assume that you wouldn't normally want to do both at the same time? or is it expected that you'd usually have a global **and** a per subgraph invalidation? If not, I think I'd prefer that we show two separate yaml examples for global and subgraph invalidations. Makes it a bit more explicit and easier to read.

File: docs/source/configuration/entity-caching.mdx
Comment (lrlna) [inline docs/source/configuration/entity-caching.mdx:137]: It would be good to explain how to choose between the two options.

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [reply docs/source/configuration/entity-caching.mdx:148]: we can have both, and per subgraph is an override of the global config.

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [reply docs/source/configuration/entity-caching.mdx:251]: that change has been merged now

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [reply docs/source/configuration/entity-caching.mdx]: done in 4cd883cbf

File: docs/source/configuration/entity-caching.mdx
Comment (Geal) [inline docs/source/configuration/entity-caching.mdx:305]:
```suggestion

#### Scalability and performance

The scalability and performance of entity cache invalidation is based on its implementation with the Redis [`SCAN` command](https://redis.io/docs/latest/commands/scan/). The `SCAN` command provides a cursor for iterating over the entire key space and returns a list of keys matching a pattern. When executing an invalidation request, the router first runs a series of `SCAN` calls and then it runs [`DEL`](https://redis.io/docs/latest/commands/del/) calls for any matching keys. 

The time complexity of a single invalidation request grows linearly with the number of entries, as each entry requires `SCAN` to iterate over. The router can also execute multiple invalidation requests simultaneously. This lowers latency but might increase the load on Redis instances.

To help tune invalidation performance and scalability, you should benchmark the ratio of the invalidation rate against the number of entries that will be recorded. If it's too low, you can tune it with the following:
- Increase the number of pooled Redis connections.
- Increasing the `SCAN` count option. This shouldn't be too large, with 1000 as a generally reasonable value, because larger values will reduce the operation throughput of the Redis instance.
- Use separate Redis instances for some subgraphs.
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:316]:
```suggestion

### Private information caching
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:148]: Adding clarification that subgraph overrides global setting:

```suggestion
You can configure entity cache invalidation globally with `preview_entity_cache.invalidation`. You can also override the global setting for a subgraph with `preview_entity_cache.subgraph.subgraphs.invalidation`. The example below shows both:
```

File: docs/source/configuration/entity-caching.mdx
Comment (shorgi) [inline docs/source/configuration/entity-caching.mdx:172]:
```suggestion
        # per subgraph invalidation configuration overrides global configuration
```

**Review by chandrikas — APPROVED** [review]

**Review by abernix — APPROVED** [review]

**Comment by Geal** [comment]
cool, this is now blocked on a failing perf test

**Comment by Geal** [comment]
finally merged

**Pull request merged**
